main/lsp/requests/rename.cc

```diff
@@ -29,47 +29,208 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
-        if (fref.data(gs).isPayload()) {
-            // We don't support renaming things in payload files.
-            return JSONNullObject();
+    vector<core::SymbolRef> subclasses;
+    for (u4 i = 1; i < gs.classAndModulesUsed(); ++i) {
+        auto s = core::SymbolRef(&gs, core::SymbolRef::Kind::ClassOrModule, i);
+        if (isSubclass(gs, root, s, memoized, visited)) {
+            subclasses.emplace_back(s);
         }
+    }
+    return subclasses;
+}
 
-        auto loc = location->range->toLoc(gs, fref);
-        auto source = loc.source(gs);
-        std::vector<std::string> strs = absl::StrSplit(source, "::");
-        strs[strs.size() - 1] = string(newName);
-        edits[location->uri].push_back(make_unique<TextEdit>(move(location->range), absl::StrJoin(strs, "::")));
+// Follow superClass links until we find the highest class that contains the given method. In other words we find the
+// "root" of the tree of classes that define a method.
+core::SymbolRef findRootClassWithMethod(const core::GlobalState &gs, core::SymbolRef klass, core::NameRef methodName) {
+    auto root = klass;
+    ENFORCE(klass.data(gs)->isClassOrModule());
+    while (true) {
+        auto tmp = root.data(gs)->superClass();
+        ENFORCE(tmp.exists()); // everything derives from Kernel::Object so we can't ever reach the actual top type
+        if (!tmp.exists() || !(tmp.data(gs)->findMember(gs, methodName).exists())) {
+            break;
+        }
+        root = tmp;
     }
+    return root;
+}
+
+class Renamer {
+public:
+    Renamer(const string oldName, const string newName) : oldName(oldName), newName(newName) {}
+    virtual ~Renamer() = default;
+    virtual void rename(const core::GlobalState &gs, unique_ptr<Location> &location, core::FileRef fref) = 0;
+    unique_ptr<WorkspaceEdit> buildEdit();
 
+protected:
+    string oldName;
+    string newName;
+    UnorderedMap<string, vector<unique_ptr<TextEdit>>> edits;
+};
+
+unique_ptr<WorkspaceEdit> Renamer::buildEdit() {
+    auto we = make_unique<WorkspaceEdit>();
     vector<unique_ptr<TextDocumentEdit>> textDocEdits;
     for (auto &item : edits) {
         // TODO: Version.
         textDocEdits.push_back(make_unique<TextDocumentEdit>(
             make_unique<VersionedTextDocumentIdentifier>(item.first, JSONNullObject()), move(item.second)));
     }
     we->documentChanges = move(textDocEdits);
-
     return we;
 }
+class MethodRenamer : public Renamer {
+public:
+    MethodRenamer(const string oldName, const string newName) : Renamer(oldName, newName) {}
+    ~MethodRenamer() {}
+    void rename(const core::GlobalState &gs, unique_ptr<Location> &location, core::FileRef fref) override {
+        auto loc = location->range->toLoc(gs, fref);
+        auto source = loc.source(gs);
+        string newsrc;
+        if (absl::StartsWith(source, "def ") || absl::StartsWith(source, "attr_reader") ||
+            absl::StartsWith(source, "attr_accessor")) {
```

Comment:
`attr_writer` is another.

We can remove this hardcoded logic, btw (maybe in the next PR). `getReferencesToSymbol` actually knows the provenance of these locations from the `QueryResponse` objects it processes and if they are from sends or definitions:

https://github.com/sorbet/sorbet/blob/e932d6f8/main/lsp/LSPTask.cc#L226

Maybe we need to introduce a new method (maybe that `getReferencesToSymbol` calls?) that returns a vector of relevant `QueryResponse` objects.